### PR TITLE
[SR GeekCamp] Distributed execution of count distinct for multiple columns 【Moderate】 #4025

### PR DESCRIPTION
**Pull request by wangxiaobaidu11**
[SR GeekCamp] Distributed execution of count distinct for multiple columns 

## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Which issues of this PR fixes ：

【Moderate】 #4025
 cte is used for implementation
for example：
select count(distinct agent_name), count(distinct team_name), count(distinct region_code) from table_v1
==>
with xx as (select agent_name, team_name, region_code from table_v1)  select * from (select count(distinct agent_name) from xx) y, (select count(distinct team_name) from xx) z , (select count(distinct region_code) from xx) m;
Performance can be doubled

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/RuleSet.java [hunk 1]
```diff
@@ -263,6 +264,13 @@ public class RuleSet {
                 new RewriteDuplicateAggregateFnRule()
         ));
 
+        rewriteRules.put(RuleSetType.MULTI_DISTINCT_REWRITE_OPTIMIZED, ImmutableList.of(
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/RuleSet.java:267]: Don't need add new `RuleSetType`

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctOptimizedRule.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+    public RewriteMultiDistinctOptimizedRule() {
+        super(RuleType.TF_REWRITE_MULTI_DISTINCT_OPTIMIZED,
+                        Pattern.create(OperatorType.LOGICAL_AGGR).addChildren(Pattern.create(
+                                        OperatorType.PATTERN_LEAF)));
+    }
+
+    @Override
+    public boolean check(OptExpression input, OptimizerContext context) {
+        LogicalAggregationOperator agg = (LogicalAggregationOperator) input.getOp();
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctOptimizedRule.java:40]: check wasn't group column

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctOptimizedRule.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+                cteId);
+
+
+        anchorOperatorInputsList.add(inputs);
+
+        context.getSessionVariable().setCboCteReuse(true);
+        context.getCteContext().setEnableCTE(true);
+        // add forced identifier
+        context.getCteContext().setForcedCTE(true);
+        context.getSessionVariable().setEnablePipelineEngine(true);
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctOptimizedRule.java:93]: check there flag in `check` method, don't execute the rule if there flag is false

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/RuleSet.java (hunk 1 above)
Comment (wangxiaobaidu11) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/RuleSet.java:267]: ok，i reuse the original RuleSetType

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 91 / 105 (86.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/CTEUtils.java|1|8|12.50%|
|:large_blue_circle:|com/starrocks/sql/optimizer/CTEContext.java|2|5|40.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|2|3|66.67%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctOptimizedRule.java|82|85|96.47%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctRule.java|3|3|100.00%|

**Review by Youngwb — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Pull request merged**